python/lsst/ctrl/mpexec/singleQuantumExecutor.py

```diff
@@ -374,6 +411,50 @@ def writeMetadata(self, quantum, metadata, taskDef, butler):
                     f" and execution") from exc
             butler.put(metadata, ref[0])
 
+    def writeLogRecords(self, quantum, taskDef, butler):
+        # If we are logging to an external file we must always try to
+        # close it.
+        filename = None
+        if isinstance(self.log_handler, FileHandler):
+            filename = self.log_handler.stream.name
+            self.log_handler.close()
+
+        if self.log_handler is not None:
+            # Remove the handler so we stop accumulating log messages.
+            logging.getLogger().removeHandler(self.log_handler)
+
+        if taskDef.logOutputDatasetName is not None and self.log_handler is not None:
+            # DatasetRef has to be in the Quantum outputs, can lookup by name
+            try:
+                ref = quantum.outputs[taskDef.logOutputDatasetName]
+            except LookupError as exc:
+                raise InvalidQuantumError(
+                    f"Quantum outputs is missing log output dataset type {taskDef.logOutputDatasetName};"
+                    f" this could happen due to inconsistent options between QuantumGraph generation"
+                    f" and execution") from exc
+
+            if isinstance(self.log_handler, ButlerLogRecordHandler):
+                butler.put(self.log_handler.records, ref[0])
+
+                # Clear the records in case the handler is reused.
+                self.log_handler.records.clear()
+            else:
+                if filename is None:
+                    raise RuntimeError("Something bad went wrong extracting filename from"
```

Comment:
`Something bad went wrong` - does it mean it's not bad anymore? 😄

Reply:
Maybe it should be an assert here because this check is never meant to trigger -- if it does something unexpected happened.